### PR TITLE
Port Stopwatch to Goodies

### PR DESCRIPTION
**Pull request by mattr555**
Porting Issue: https://github.com/duckduckgo/zeroclickinfo-spice/issues/2463

Deleting PR: https://github.com/duckduckgo/zeroclickinfo-spice/pull/2665

https://duck.co/ia/view/goodie

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/Stopwatch.pm [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+triggers startend => 'stopwatch', 'stop watch';
+
+# Handle statement
+handle remainder => sub {
+    return unless ($_ eq '' || $_ eq 'online');
+
+    return "",
+        structured_answer => {
+            id => 'stopwatch',
```
Comment (moollaza) [inline lib/DDG/Goodie/Stopwatch.pm]: @mattr555 we've actually removed the `id` and `name` from Goodies now. They get them from the metadata :+1:

**Comment by moollaza** [comment]
@mattr555 for now you'll need a new IA page so we can get this live. We can't change the live IA page to Goodie so create a stopwartch_goodie page and link that here. Then we can put this on beta and the Spice removal PR. Once merged we'll remove the goodie page, or change it's ID

**Comment by mattr555** [comment]
@moollaza IA page created! However, now that I've removed the id and name attributes from the `structured_answer` the answer doesn't trigger on duckpan. Is this expected?

**Comment by mintsoft** [comment]
@mattr555 It triggers for me?
```
07:55:11 ~/scripts/ddg/ddh/goodies (mattr/stopwatch-port)
rob@set $ duckpan query
Checking for DuckPAN requirements...
Checking perl version...
Checking for latest App::DuckPAN...
[NOTICE] You have version 0.204, latest is 1000!
[NOTICE] Please install the latest App::DuckPAN package with: duckpan upgrade
installing DDG at /home/rob/perl5/lib/perl5/App/DuckPAN.pm line 568.
Checking for latest DDG Perl package...
[NOTICE] You have version 0.167, latest is 1000!
[NOTICE] Please install the latest DDG package with: duckpan DDG
Checking for ssh...
Checking for git...
Loading Instant Answers...
(Empty query for ending test)
Query: online stopwatch
  You entered: online stopwatch
---
DDG::ZeroClickInfo  {
    Parents       WWW::DuckDuckGo::ZeroClickInfo
    public methods (4) : DOES, has_structured_answer, new, structured_answer
    private methods (0)
    internals: {
        answer              "",
        answer_type         "stopwatch",
        caller              "DDG::Goodie::Stopwatch",
        is_cached           1,
        is_unsafe           0,
        structured_answer   {
            data        {},
            meta        {
                itemType     "stopwatch",
                sourceName   "Stopwatch"
            },
            templates   {
                detail        "DDH.stopwatch.stopwatch",
                group         "base",
                wrap_detail   "base_detail"
            }
        }
    }
}
---
```

**Comment by mattr555** [comment]
@mintsoft sorry, should have been more clear: it triggers, but the JavaScript doesn't run on the page.

**Comment by zachthompson** [comment]
It's the new ID mismatching with everything else that is still "stopwatch".  If you replace `$caller` [here](https://github.com/duckduckgo/p5-app-duckpan/blob/master/lib/App/DuckPAN/Web.pm#L307) with `'DDG::Spice::Stopwatch'` it works fine.

@moollaza Probably best to just reuse the current IA page and update the values, i.e. type, blockgroup, Perl module, etc., once released and ready to swap.

**Comment by mattr555** [comment]
@zachthompson Hmm...I think I'm missing something. When I set `zci caller => 'DDG::Spice::Stopwatch'` I get `Template not found: DDH.stopwatch.stopwatch`. Same if I then set `structured_answer->templates->detail` to `'Spice.stopwatch.stopwatch'`.

**Comment by zachthompson** [comment]
@mattr555 Not quite sure where you are setting that value.  However, try updating App/DuckPAN/Web.pm as in the screenshot.

stopwatch

**Comment by mattr555** [comment]
I changed all instances of the id in the JS and CSS to `stopwatch_goodie` and that seems to fix everything. @zachthompson I'm not sure if this is how you want me to proceed? Can we continue like this and then just delete/deprecate the old IA page?

**Comment by mintsoft** [comment]
@mattr555 I think that's the plan; we might as well get the bulk of the work dealt with

**Comment by moollaza** [comment]
Deploying to Beta...

**Comment by mattr555** [comment]
@moollaza Can we disable the old spice on beta without affecting the production site? I think the code is running for both and interfering with each other. (Also it seems like the css with the right `.zci--stopwatch_goodie` classes hasn't loaded)

**Comment by matthewmorris085** [comment]
I think if your talking about what I think your talking about its fine I believe if you run the program again but have a interface between it it will run good depending on device

**Comment by moollaza** [comment]
> Can we disable the old spice on beta without affecting the production site?

@mattr555 not with the metadata, but I just noticed the Spice removal wasn't deployed to Beta so I'm doing that now. That should leave us with only the Goodie on Beta

**Comment by moollaza** [comment]
@mattr555 the Goodie is now the only thing on Beta. It looks like when you start and stop the timer, we get two identical rows. In production we only show one though. Is this a bug?

**Comment by moollaza** [comment]
stopwatch_at_duckduckgo

**Comment by moollaza** [comment]
@mattr555 OK I've done some talking with Zach and I think we have this figured out. We need to update the code here as if the ID was `stopwatch` instead of `stopwatch__goodie`.

Then I can merge this and during the next deploy update the IA Page to be a Goodie, which will remove the Spice code from being used. Then we're free to remove the Spice code.

**Comment by moollaza** [comment]
Going to Merge + Release this tomorrow..

/cc @MariagraziaAlastra

**Comment by moollaza** [comment]
This LGTM :+1:

**Comment by moollaza** [comment]
Everything went smoothly. It will go live later today, you shouldn't notice anything change really ;)

https://moollaza.duckduckgo.com/?q=stopwatch&ia=answer

**Comment by moollaza** [comment]
The frontend/backend files broke, but we'll be fixing

**Comment by mattr555** [comment]
Thanks a lot for the help! 👍

**Pull request merged**